File: ai-daily/scripts/tts_split_voice_to_audio.py
#!/usr/bin/env python3
"""
把“口播文字稿”拆成开场/每条新闻/收尾，并分别调用 voice_to_audio.py 生成 MP3。

为什么要拆：
- Remotion 需要 openingAudio + 每条 news.audio + endingAudio（而不是一整个 mp3）。

输入：
- voice script md（建议结构类似：
  # xxx
  ## 开场白
  ...
  ## 正文
  ### 第一条
  ...
  ### 第二条
  ...
  ## 收尾
  ...
）

输出：
- out-dir/
  - opening.mp3
  - news-1.mp3, news-2.mp3, ...
  - ending.mp3
"""

from __future__ import annotations

import argparse
from pathlib import Path
import re
import subprocess
import sys
import tempfile
from typing import Dict, List, Tuple


def _read_text(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def _write_text(path: Path, content: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")


def _extract_section(md: str, header: str) -> str | None:
    """
    提取以 `## <header>` 开头的段落，直到下一个 `## ` 同级标题或文末。
    """
    # header 形如 "开场白" / "正文" / "收尾"
    m = re.search(rf"^##\s+{re.escape(header)}\s*$", md, flags=re.MULTILINE)
    if not m:
        return None
    start = m.end()
    m2 = re.search(r"^##\s+.+$", md[start:], flags=re.MULTILINE)
    end = start + (m2.start() if m2 else len(md) - start)
    return md[start:end].strip()


def _extract_body_news(md: str) -> List[str]:
    """
    从 `## 正文` 内按 `### ...` 拆成多条文本段。
    返回每条（去掉标题行本身）。
    """
    body = _extract_section(md, "正文")
    if not body:
        return []

    # 用 `###` 分段；标题行不参与 TTS 文本
    parts = re.split(r"^###\s+.+$", body, flags=re.MULTILINE)
    # split 会丢失标题，因此需要同时剔除空段
    segments = [p.strip() for p in parts if p.strip()]
    return segments


def _prepare_tmp_md(tmp_dir: Path, title: str, text: str) -> Path:
    p = tmp_dir / f"{title}.md"
    # voice_to_audio --plain 会丢弃首行标题；我们只放正文，避免 TTS 读出“第一条”
    _write_text(p, f"# {title}\n\n{text.strip()}\n")
    return p


def run_voice_to_audio(voice_md: Path, out_mp3: Path, pause_ms: int | None = None) -> None:
    script = Path(__file__).parent / "voice_to_audio.py"
    cmd: List[str] = [sys.executable, str(script), str(voice_md), "-o", str(out_mp3), "--plain"]
    if pause_ms is not None:
        # --plain 禁用 SSML；pause-ms 对 pure text 可能无意义，但保留接口可控
        cmd.extend(["--pause-ms", str(pause_ms)])

    r = subprocess.run(cmd, capture_output=True, text=True)
    if r.returncode != 0:
        raise RuntimeError(f"voice_to_audio failed: {r.stderr.strip() or r.stdout.strip()}")


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("voice_md", help="口播文字稿 .md 路径（含 开场白/正文/收尾）")
    ap.add_argument("--out-dir", required=True, help="输出目录（存放 opening/news-*/ending.mp3）")
    ap.add_argument("--pause-ms", type=int, default=None, help="可选：兼容参数（在 --plain 模式下通常无效）")
    args = ap.parse_args()

    voice_md_path = Path(args.voice_md).expanduser().resolve()
    out_dir = Path(args.out_dir).expanduser().resolve()
    if not voice_md_path.exists():
        raise SystemExit(f"voice md not found: {voice_md_path}")

    md = _read_text(voice_md_path)
    opening = _extract_section(md, "开场白") or ""
    ending = _extract_section(md, "收尾") or ""
    news_segments = _extract_body_news(md)

    if not opening.strip():
        print("[Warn] 开场白段落为空或缺失，将仍尝试合成空文本（可能失败）。", file=sys.stderr)
    if not news_segments:
        raise SystemExit("未能从 `## 正文` 中解析到任意 `###` 新闻分段。请检查模板。")
    if not ending.strip():
        print("[Warn] 收尾段落为空或缺失，将仍尝试合成空文本（可能失败）。", file=sys.stderr)

    out_dir.mkdir(parents=True, exist_ok=True)
    opening_mp3 = out_dir / "opening.mp3"
    ending_mp3 = out_dir / "ending.mp3"

    with tempfile.TemporaryDirectory(prefix="ai-daily-tts-") as td:
        tmp_dir = Path(td)
        # Opening
        opening_tmp = _prepare_tmp_md(tmp_dir, "opening", opening)
        print(f"[TTS] opening -> {opening_mp3}")
        run_voice_to_audio(opening_tmp, opening_mp3, pause_ms=args.pause_ms)

        # News
        for i, seg in enumerate(news_segments, start=1):
            news_tmp = _prepare_tmp_md(tmp_dir, f"news-{i}", seg)
            mp3p = out_dir / f"news-{i}.mp3"
            print(f"[TTS] news {i} -> {mp3p}")
            run_voice_to_audio(news_tmp, mp3p, pause_ms=args.pause_ms)

        # Ending
        ending_tmp = _prepare_tmp_md(tmp_dir, "ending", ending)
        print(f"[TTS] ending -> {ending_mp3}")
        run_voice_to_audio(ending_tmp, ending_mp3, pause_ms=args.pause_ms)


if __name__ == "__main__":
    main()

File: ai-daily/scripts/tts_split_voice_to_audio.ts
#!/usr/bin/env bun
// @ts-nocheck

import { existsSync, mkdtempSync, readFileSync, rmSync, mkdirSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { resolve } from "node:path";
import { execaSync } from "execa";
import { Command } from "commander";
import { z } from "zod";

const CliOptionsSchema = z.object({
  outDir: z.string().min(1),
  pauseMs: z.string().regex(/^\d+$/).optional(),
  reportMd: z.string().min(1).optional(),
});

function extractSection(md: string, header: string): string | null {
  const m = md.match(new RegExp(`^##\\s+${header}\\s*$`, "m"));
  if (!m || m.index == null) return null;
  const start = m.index + m[0].length;
  const rest = md.slice(start);
  const m2 = rest.match(/^##\s+.+$/m);
  const end = start + (m2?.index ?? rest.length);
  return md.slice(start, end).trim();
}

function extractBodyNews(md: string): string[] {
  const body = extractSection(md, "正文");
  if (!body) return [];
  return body
    .split(/^###\s+.+$/gm)
    .map((x) => x.trim())
    .filter(Boolean);
}

function parseTodayNewsCount(reportMd: string): number {
  const h = reportMd.match(/^##\s+今日资讯\s*$/m);
  if (!h || h.index == null) return 0;
  const start = h.index + h[0].length;
  const rest = reportMd.slice(start);
  const next = rest.match(/^##\s+.+$/m);
  const block = reportMd.slice(start, start + (next?.index ?? rest.length)).trim();
  const lines = block
    .split("\n")
    .map((x) => x.trim())
    .filter((x) => /^\d+\.\s/.test(x));
  return lines.length;
}

function runVoiceToAudio(voiceMd: string, outMp3: string, pauseMs?: string): void {
  const script = resolve(import.meta.dir, "voice_to_audio.ts");
  const cmd = ["bun", script, voiceMd, "-o", outMp3, "--plain"];
  if (pauseMs) cmd.push("--pause-ms", pauseMs);
  const r = execaSync(cmd[0], cmd.slice(1), { reject: false });
  if (r.exitCode !== 0) {
    const msg = [String(r.stdout ?? "").trim(), String(r.stderr ?? "").trim()].filter(Boolean).join("\n");
    throw new Error(msg || `voice_to_audio failed: exitCode=${r.exitCode}`);
  }
}

function estimateDurationSec(text: string): number {
  const chars = text.trim().length;
  return Math.max(2, Math.ceil(chars / 14));
}

function makeSilentMp3(pathValue: string, durationSec: number): void {
  execaSync("ffmpeg", [
    "-y",
    "-f",
    "lavfi",
    "-i",
    "anullsrc=r=24000:cl=mono",
    "-t",
    String(durationSec),
    "-q:a",
    "9",
    "-acodec",
    "libmp3lame",
    pathValue,
  ]);
}

function runWithFallback(voiceMd: string, outMp3: string, sourceText: string, pauseMs?: string): void {
  try {
    runVoiceToAudio(voiceMd, outMp3, pauseMs);
  } catch (e) {
    const msg = String(e instanceof Error ? e.message : e);
    if (msg.includes("quota exceeded")) {
      const sec = estimateDurationSec(sourceText);
      console.error(`[TTS][fallback] 检测到配额不足，改为静音占位音频: ${outMp3} (${sec}s)`);
      makeSilentMp3(outMp3, sec);
      return;
    }
    throw e;
  }
}

function main(): void {
  const program = new Command();
  program
    .argument("<voice-md>")
    .requiredOption("--out-dir <dir>")
    .option("--pause-ms <ms>")
    .option("--report-md <path>");
  program.parse(process.argv);
  const options = CliOptionsSchema.parse(program.opts());
  const voiceArg = z.string().min(1).parse(program.args[0]);
  const voiceMdPath = resolve(String(voiceArg));
  if (!existsSync(voiceMdPath)) throw new Error(`voice md not found: ${voiceMdPath}`);
  const outDir = resolve(String(options.outDir));
  const pauseMs = options.pauseMs ? String(options.pauseMs) : "";
  const reportMdPath = options.reportMd ? resolve(String(options.reportMd)) : "";

  const md = readFileSync(voiceMdPath, "utf-8");
  const opening = extractSection(md, "开场白") ?? "";
  const ending = extractSection(md, "收尾") ?? "";
  const newsSegments = extractBodyNews(md);
  const expectedNewsCount =
    reportMdPath && existsSync(reportMdPath) ? parseTodayNewsCount(readFileSync(reportMdPath, "utf-8")) : 0;

  if (!newsSegments.length) throw new Error("未能从 `## 正文` 解析到任何 `###` 新闻分段。");
  if (!opening.trim()) console.error("[Warn] 开场白段落为空或缺失。");
  if (!ending.trim()) console.error("[Warn] 收尾段落为空或缺失。");

  mkdirSync(outDir, { recursive: true });
  const tempDir = mkdtempSync(resolve(tmpdir(), "ai-daily-tts-"));
  const tmpMd = (title: string, text: string): string => {
    const p = resolve(tempDir, `${title}.md`);
    writeFileSync(p, `# ${title}\n\n${text.trim()}\n`, "utf-8");
    return p;
  };

  try {
    console.log(`[TTS] opening -> ${resolve(outDir, "opening.mp3")}`);
    runWithFallback(tmpMd("opening", opening), resolve(outDir, "opening.mp3"), opening, pauseMs || undefined);
    for (let i = 0; i < newsSegments.length; i++) {
      const out = resolve(outDir, `news-${i + 1}.mp3`);
      console.log(`[TTS] news ${i + 1} -> ${out}`);
      runWithFallback(tmpMd(`news-${i + 1}`, newsSegments[i]), out, newsSegments[i], pauseMs || undefined);
    }
    if (expectedNewsCount > newsSegments.length) {
      for (let i = newsSegments.length; i < expectedNewsCount; i++) {
        const out = resolve(outDir, `news-${i + 1}.mp3`);
        const placeholder = `第${i + 1}条新闻音频占位`;
        console.error(`[TTS][fallback] 语音脚本条数不足，补齐静音: ${out}`);
        makeSilentMp3(out, estimateDurationSec(placeholder));
      }
    }
    console.log(`[TTS] ending -> ${resolve(outDir, "ending.mp3")}`);
    runWithFallback(tmpMd("ending", ending), resolve(outDir, "ending.mp3"), ending, pauseMs || undefined);
  } finally {
    rmSync(tempDir, { recursive: true, force: true });
  }
}

try {
  main();
} catch (e) {
  console.error(String(e instanceof Error ? e.message : e));
  process.exit(1);
}

